### ISSUE TITLE
Using dynamic=True in a keras Metric constructor should run the metric methods in eager mode

### ISSUE THREAD

**Issue by jvishnuvardhan** [keras-team/keras#15758]
Moving this feature request from TF repository https://github.com/tensorflow/tensorflow/issues/39114
Currently, when using dynamic=True in a Metric constructor, it has no effect. All the methods are run in graph mode. Since this feature is already present in keras Layers, and that a Metric is a subclass of Layer, one would expect a metric to run in eager mode if dynamic=True is provided.

Will this change the current api? How?

No change to the API. dynamic=True is already accepted in a Metric constructor. The change would be in behavior only. The flag should trigger the eager mode. Currently it doesn't do anything.

Who will benefit with this feature?

Users who want to make custom metrics which need eager mode and want to distribute them.

Any Other info.

Here is a minimal notebook reproducing the problem. With this feature implemented, the error should not happen:

https://colab.research.google.com/drive/11EuSPoyedceHySz_uxxqjZJbU2uatpfk

**Comment by haifeng-jin** [comment]
Triage notes:

If you want the metrics to run eagerly, you can just use `tf.config.run_functions_eagerly` to run everything eagerly.

**Comment by foxik** [comment]
BTW, one can also explicitly use a `tf.numpy_function` / `tf.py_function` in the metric to compute it using a Python code.
